src/common/browser-adapters/browser-adapter-event-manager.ts

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import { PromiseFactory } from 'common/promises/promise-factory';
+import { TimeoutFactory, TimeoutType } from 'common/timeouts/timeout-factory';
+import { remove } from 'lodash';
+import { DictionaryStringTo } from 'types/common-types';
+import { Events } from 'webextension-polyfill';
+export interface ApplicationListener {
+    (...args: any[]): any;
+}
+
+export interface EventDetails {
+    eventType: string;
+    eventArgs: any[];
+}
+
+export interface AdapterListener {
+    (eventType: string): (eventArgs: any[]) => any;
+}
+
+const TWO_MINUTES = 120000;
+const FOUR_MINUTES = 2 * TWO_MINUTES;
+
+export class BrowserAdapterEventManager {
+    protected deferredEvents: EventDetails[] = [];
+    protected eventsToApplicationListenersMapping: DictionaryStringTo<ApplicationListener[]> = {};
+    protected adapterListener: AdapterListener =
+        (eventType: string) =>
+        (...eventArgs: any[]) => {
+            this.processEvent(eventType, eventArgs);
```

Comment:
If this is intended to be the thing that gets attached to browser events directly, it needs to be returning the promise generated by `forwardEventToApplicationListener`, no?

Reply:
(and probably creating some sort of deferred-with-timeout Promise in the event that there are no current app listeners)